Overhaul external process calls

tomspur opened the pull request:
This pull request starts to implement the move from `os.system` to the `subprocess` module and captures stderr and includes it in the exceptions.

For more information see #7490.

QuLogic commented:
I wonder if we should have been using our `matplotlib.compat.subprocess` for compatibility with GAE, or whether it makes no difference because you'd never have LaTeX installed there (the other instances of `subprocess` you've changed here don't really matter for this, I think.)

tomspur commented:
Thanks for the pointer to `matplotlib.compat.subprocess`. I think using it for consistency is also fine and I'll change it to it.

Just as a reference of the benefit of this:
With the failing outputs of TexManager, I see now in a more detail why some tests failed
```
ERROR: test suite for <class 'matplotlib.sphinxext.tests.test_tinypages.TestTinyPages'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tomspur/.virtualenvs/default_python3/lib/python3.5/site-packages/nose/suite.py", line 210, in run
    self.setUp()
  File "/home/tomspur/.virtualenvs/default_python3/lib/python3.5/site-packages/nose/suite.py", line 293, in setUp
    self.setupContext(ancestor)
  File "/home/tomspur/.virtualenvs/default_python3/lib/python3.5/site-packages/nose/suite.py", line 316, in setupContext
    try_run(context, names)
  File "/home/tomspur/.virtualenvs/default_python3/lib/python3.5/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/sphinxext/tests/test_tinypages.py", line 56, in setup_class
    raise e
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/sphinxext/tests/test_tinypages.py", line 53, in setup_class
    '{0}\nstderr:\n{1}\n'.format(out, err))
RuntimeError: sphinx-build failed with stdout:
b'Running Sphinx v1.4.8\nmaking output directory...\n'
stderr:
b'\nException occurred:\n  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/rcsetup.py", line 847, in validate_cycler\n    cycler_inst.change_key(prop, norm_prop)\nAttributeError: \'Cycler\' object has no attribute \'change_key\'\nThe full traceback has been saved in /tmp/sphinx-err-0zxJFz.log, if you want to report the issue to the developers.\nPlease also report this if it was a user error, so that a better error message can be provided next time.\nA bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!\n'
```

tomspur commented:
Having access to the stderr seems to work fine:
```
ERROR: matplotlib.tests.test_backend_pgf.test_rcupdate
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tomspur/.virtualenvs/default_python3/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/testing/decorators.py", line 150, in wrapped_callable
    func(*args, **kwargs)
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/testing/decorators.py", line 479, in backend_switcher
    result = func(*args, **kwargs)
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/tests/test_backend_pgf.py", line 144, in test_rcupdate
    compare_figure('pgf_rcupdate%d.pdf' % (i + 1), tol=tol[i])
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/tests/test_backend_pgf.py", line 43, in compare_figure
    plt.savefig(actual, **savefig_kwargs)
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/pyplot.py", line 696, in savefig
    res = fig.savefig(*args, **kwargs)
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/figure.py", line 1694, in savefig
    self.canvas.print_figure(*args, **kwargs)
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/backend_bases.py", line 2230, in print_figure
    **kwargs)
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/backends/backend_pgf.py", line 929, in print_pdf
    self._print_pdf_to_fh(fh, *args, **kwargs)
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/backends/backend_pgf.py", line 883, in _print_pdf_to_fh
    self.print_pgf(fname_pgf, *args, **kwargs)
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/backends/backend_pgf.py", line 865, in print_pgf
    self._print_pgf_to_fh(fh, *args, **kwargs)
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/backends/backend_pgf.py", line 844, in _print_pgf_to_fh
    RendererPgf(self.figure, fh),
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/backends/backend_pgf.py", line 422, in __init__
    self.latexManager = LatexManagerFactory.get_latex_manager()
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/backends/backend_pgf.py", line 234, in get_latex_manager
    new_inst = LatexManager()
  File "/home/tomspur/src/matplotlib.git/build/lib.linux-x86_64-3.5/matplotlib/backends/backend_pgf.py", line 318, in __init__
    raise LatexError("LaTeX returned an error, probably missing font or error in preamble:\n%s" % stdout)
matplotlib.backends.backend_pgf.LatexError: LaTeX returned an error, probably missing font or error in preamble:
b"This is pdfTeX, Version 3.14159265-2.6-1.40.17 (TeX Live 2016) (preloaded format=pdflatex)\n restricted \\write18 enabled.\n**entering extended mode\nLaTeX2e <2016/03/31>\nBabel <3.9r> and hyphenation patterns for 6 language(s) loaded.\n\n*(/usr/share/texlive/texmf-dist/tex/latex/base/minimal.cls\nDocument Class: minimal 2001/05/25 Standard LaTeX minimal class\n)\n*(/usr/share/texlive/texmf-dist/tex/latex/base/inputenc.sty\n(/usr/share/texlive/texmf-dist/tex/latex/ucs/utf8x.def))\n(/usr/share/texlive/texmf-dist/tex/latex/ucs/ucs.sty\n(/usr/share/texlive/texmf-dist/tex/latex/ucs/data/uni-global.def))\n*(/usr/share/texlive/texmf-dist/tex/latex/base/fontenc.sty\n(/usr/share/texlive/texmf-dist/tex/latex/base/t1enc.def))\n*\n! LaTeX Error: File `sfmath.sty' not found.\n\nType X to quit or <RETURN> to proceed,\nor enter new name. (Default extension: sty)\n\nEnter file name: ! Undefined control sequence.\n\\filename@simple ...#2\\\\}\\fi \\edef \\filename@base \n                                                  {#1}\n<*> \n    \n!  ==> Fatal error occurred, no output PDF file produced!\nTranscript written on texput.log.\n"

```

Which points to `LaTeX Error: File sfmath.sty not found.`

tomspur commented:
The failure seems to be related to an upgrade of freetype on travis...

QuLogic commented:
No, it's just a flaky test; we build in a specific freetype version on CI.

QuLogic commented:
Can you rebase and remove the use of `shell=True`?

tomspur commented:
Removing `shell=True` turns out to require a rewrite of most of TexManager as there are chained commands run in the subprocess. So this will likely still take a while.

QuLogic commented:
It's mostly just `cd` which can be replaced by passing `cwd=` and outout redirection which can be fixed using the right `stdout=`.

tomspur commented:
Thanks for the pointer. I wasn't aware of the `stdout=` parameter.

tomspur commented:
There are now only two `shell=True` left, which are windows specific and I cannot test them properly here on Linux...

tacaswell commented:
You should always import from our compat shim, https://github.com/matplotlib/matplotlib/blob/1a9b4eecff4409b4d9e5502df2ab12b4b16ac1f8/lib/matplotlib/compat/subprocess.py due to issues on early micro versions of 2.7 on mac (see https://github.com/matplotlib/matplotlib/issues/5314 )

tacaswell commented on examples/tests/backend_driver.py:
```diff
@@ -350,7 +350,7 @@ def run(arglist):
 except ImportError:
     def run(arglist):
-        os.system(' '.join(arglist))
+        os.system(arglist)
```
Can just drop this whole section, this is <2.7 compat code.

tomspur replied:
Removed this part of the backend driver and used the compat shim also there.

tacaswell commented:
Ah, disregard my comment, I see you are using the shim internally but not in the examples.

Carry on :sheep:

tomspur commented:
I don't know how to fix the python 2.7 failure. It looks quite similar to #7715, which @tacaswell has tracked down. Any pointers how to fix it here?

jkseppan commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1499,30 +1504,28 @@ def gs_distill(tmpfile, eps=False, ptype='letter', bbox=None, rotated=False):
-                "%s" > "%s"'% (gs_exe, dpi, device_name,
-                               paper_option, psfile, tmpfile, outfile)
-
+    command = [gs_exe, "-dBATCH", "-dNOPAUSE", "-r%d" % dpi,
```
This probably needs to be `str(gs_exe)`

jkseppan commented:
The fix to #7715 is #7804, and it added a non-ASCII directory on the path to catch this kind of bugs. The fix was to wrap the arguments to `subprocess.check_output` in `str(...)` so probably you need to do something similar here.

tomspur commented:
Thanks @jkseppan, this instance and many more were it and it works now also on python 2.7

QuLogic commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -83,8 +84,7 @@ def gs_version(self):
         from matplotlib.compat.subprocess import Popen, PIPE
-        s = Popen(self.gs_exe + " --version",
-                     shell=True, stdout=PIPE)
+        s = Popen([str(self.gs_exe), " --version"], stdout=PIPE)
```
I believe the extra space here is breaking this version check thus causing your Windows failures.

tomspur commented:
Thanks @QuLogic. The tests pass now also on Windows.

QuLogic commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1499,30 +1504,28 @@ def gs_distill(tmpfile, eps=False, ptype='letter', bbox=None, rotated=False):
-        else:
+    try:
+        with open(outfile, "w") as fout:
+            subprocess.check_call(command,
```
Can you not just use `check_output`? Isn't the `outfile` deleted right after this?

tomspur replied:
In that case the output of check_output is not defined in the `except` clause, e.g. in:
```import subprocess
try:
    output = subprocess.check_call("ls asdfasdf")
except:
    print("FAIL:", output)
```

--> `NameError: name 'output' is not defined`

dopplershift replied:
That's because in the case of an exception, nothing is actually assigned to `output`.

QuLogic replied:
For the exception, you can use `exc.output` like [here](https://github.com/matplotlib/matplotlib/pull/7572/files#diff-d593c407e3c1e32bc40e3d613669377cR411) and other places.

QuLogic commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1555,33 +1558,53 @@ def xpdf_distill(tmpfile, eps=False, ptype='letter', bbox=None, rotated=False):
+
+    try:
+        with open(outfile, "w") as fout:
+            subprocess.check_call(command,
```
Same thing here, I guess.

QuLogic commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1555,33 +1558,53 @@ def xpdf_distill(tmpfile, eps=False, ptype='letter', bbox=None, rotated=False):
+    verbose.report(command, 'debug')
+    try:
+        with open(outfile, "w") as fout:
+            subprocess.check_call(command,
```
Also here.

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -43,6 +43,7 @@
 import glob
 import os
 import shutil
+from matplotlib.compat.subprocess import subprocess
```
Move down to `matplotlib` section.

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -398,30 +399,34 @@ def make_dvi(self, tex, fontsize):
-                'latex -interaction=nonstopmode %s > "%s"' %
-                (os.path.split(texfile)[-1], outfile))
+            command = [str("latex"), "-interaction=nonstopmode",
+                       os.path.split(texfile)[-1]]
```
`os.path.split(texfile)`

dopplershift replied:
Don'you mean `os.path.basename(texfile)`?

QuLogic replied:
Err, yep.

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -398,30 +399,34 @@ def make_dvi(self, tex, fontsize):
-            else:
+                try:
+                    with open(outfile, "w") as fout:
+                        subprocess.check_call(command,
```
This can be changed to `check_output` as well, I think.

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -398,30 +399,34 @@ def make_dvi(self, tex, fontsize):
+                        ('LaTeX was not able to process the following '
+                         'string:\n%s\nLaTeX failed with: %s\n'
+                         'Here is the full report generated by LaTeX: '
+                         '\n\n' % (exc.output,
```
`exc.output` should be empty because you sent output to `fout` but you can use `report` once switched to `check_output`.

tomspur replied:
I removed `exc.output`, but I don't think it is possible to switch to `check_output` (see above the other comment)

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -452,25 +457,34 @@ def make_dvi_preview(self, tex, fontsize):
-                'latex -interaction=nonstopmode %s > "%s"' %
-                (os.path.split(texfile)[-1], outfile))
+            command = [str("latex"), "-interaction=nonstopmode",
+                       os.path.split(texfile)[-1]]
```
`os.path.basename`

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -452,25 +457,34 @@ def make_dvi_preview(self, tex, fontsize):
-            exit_status = os.system(command)
+            try:
+                with open(outfile, "w") as fout:
+                    subprocess.check_call(command,
```
Probably can be `check_output` as well.

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -452,25 +457,34 @@ def make_dvi_preview(self, tex, fontsize):
+                    subprocess.check_call(command,
+                                          cwd=self.texcache,
+                                          stdout=fout,
+                                         stderr=subprocess.STDOUT)
```
Alignment is off by one space.

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -506,25 +520,35 @@ def make_png(self, tex, fontsize, dpi):
-                os.path.split(dvifile)[-1], outfile))
+            command = [str("dvipng"), "-bg", "Transparent", "-D", str(dpi),
+                       "-T", "tight", "-o", os.path.split(pngfile)[-1],
+                       os.path.split(dvifile)[-1]]
```
`os.path.basename` for both of these.

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -506,25 +520,35 @@ def make_png(self, tex, fontsize, dpi):
-            exit_status = os.system(command)
+            try:
+                with open(outfile, "w") as fout:
+                    subprocess.check_call(command,
```
`check_output` as well, I think.

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -544,21 +568,25 @@ def make_ps(self, tex, fontsize):
-                (os.path.split(psfile)[-1],
-                 os.path.split(dvifile)[-1], outfile))
+            command = [str("dvips"), "-q", "-E", "-o", os.path.split(psfile)[-1],
+                       os.path.split(dvifile)[-1]]
```
`os.path.basename` for both.

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -544,21 +568,25 @@ def make_ps(self, tex, fontsize):
-                if exit_status:
+            try:
+                with open(outfile, "w") as fout:
+                    subprocess.check_call(command,
```
Another `check_output`.

QuLogic commented on tools/subset.py:
```diff
@@ -145,7 +146,7 @@ def subset_font_raw(font_in, font_out, unicodes, opts):
         print('Generate("' + font_out + '")', file=pe)
         pe.close()
-        os.system("fontforge -script " + pe_fn)
+        subprocess.call("fontforge -script " + pe_fn, shell=True)
```
Make this a list and remove `shell=True` like the others (`pe_fn` is just a single file name).

QuLogic commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1440,27 +1440,32 @@ def convert_psfrags(tmpfile, psfrags, font_preamble, custom_preamble,
     command = '%s cd "%s" && latex -interaction=nonstopmode "%s" > "%s"'\
```
This should be converted into a list and given the same treatment as the `backends_ps.py` file using the `cwd` argument.

tomspur replied:
I didn't want to try it as I couldn't test the `precmd` part on Linux.
Is that taken care of automatically with using `cmd` or would you still need the `precmd` somewhere else still?

QuLogic replied:
Oops, I meant to say, do the same thing as `texmanager.py`; `precmd` is just setting the drive letter, which is the same thing that `_get_shell_cmd` did there, and it seems to work on AppVeyor.

tomspur replied:
Hmm, I don't see where `_get_shell_cmd` is used at all in the whole matplotlib. I don't see how to use that here to change this command string to a list.

QuLogic replied:
You've already removed it (e.g., [here](https://github.com/matplotlib/matplotlib/pull/7572/files#diff-d593c407e3c1e32bc40e3d613669377cL401)).

tomspur replied:
Oh, thanks...

QuLogic commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1440,27 +1440,32 @@ def convert_psfrags(tmpfile, psfrags, font_preamble, custom_preamble,
+    except subprocess.CalledProcessError as exc:
+        with io.open(outfile, 'rb') as fh:
             raise RuntimeError('LaTeX was not able to process your file:\
-    \nHere is the full report generated by LaTeX: \n\n%s'% fh.read())
```
This is (and was before) pretty ugly; please take advantage of automatic string concatenation for this:
```python
     raise RuntimeError('A string '
                        'continued on another line')
```

tomspur commented on examples/pylab_examples/movie_demo.py:
```diff
@@ -22,8 +22,10 @@
 print('Making movie animation.mpg - this may take a while')
-os.system("mencoder 'mf://_tmp*.png' -mf type=png:fps=10 -ovc lavc -lavcopts vcodec=wmv2 -oac copy -o animation.mpg")
-#os.system("convert _tmp*.png animation.mng")
+subprocess.call(["mencoder", "'mf://_tmp*.png'", "-mf", "type=png:fps=10",
```
I think that with the shell, this is more readable, so I just reverted this part (and deleted the next line)

tomspur commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -83,8 +84,7 @@ def gs_version(self):
         from matplotlib.compat.subprocess import Popen, PIPE
-        s = Popen(self.gs_exe + " --version",
-                     shell=True, stdout=PIPE)
+        s = Popen([str(self.gs_exe), "--version"], stdout=PIPE)
```
Indeed

QuLogic replied:
One missed change?

QuLogic commented:
Ping @tomspur? You can ignore the build errors; they're just flaky tests.

tomspur commented:
I just rebased it to master and cleaned up some parts. Let's see how it looks like now.

Things not implemented yet:
- converting of a string with `precmd` to a list: I don't have a windows to test this on and the example of `_get_shell_cmd` in TexManager is never used actually in the repository.
- use of `check_output` in try except: When the command fails, nothing is written to `output` and we would have no information what went wrong in the subprocess.

tomspur commented:
All comments should be implemented - the failing test in `test_invisible_Line_rendering` seems a false positive.

QuLogic commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1643,14 +1641,13 @@ def get_bbox_header(lbrt, rotated=False):
-                (gs_exe, tmpfile)
+    command = '%s -dBATCH -dNOPAUSE -sDEVICE=bbox "%s"' % (gs_exe, tmpfile)
     verbose.report(command, 'debug')
     stdin, stdout, stderr = os.popen3(command)
```
This one could use `subprocess.Popen`, probably.

tomspur replied:
I missed that... Thanks

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -451,30 +429,25 @@ def make_dvi_preview(self, tex, fontsize):
-                mpl.verbose.report(report, 'debug')
+                     'string:\n%s\n\n'
+                     'Here is the full report generated by LaTeX:\n%s '
+                     '\n\n' % (repr(tex.encode('unicode_escape')),
```
Did something require adding the unicode escaping? Windows?

tomspur replied:
Good question. This was like this in one of the calls the current master and I took that over in all new `subprocess.check_output`s

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -505,30 +478,22 @@ def make_png(self, tex, fontsize, dpi):
+                    ('dvipng was not able to process the following '
+                     'string:\n%s\n\n'
+                     'Here is the full report generated by dvipng:\n%s '
+                     '\n\n' % (repr(tex.encode('unicode_escape')),
```
Same question?

QuLogic commented on lib/matplotlib/texmanager.py:
```diff
@@ -543,23 +508,22 @@ def make_ps(self, tex, fontsize):
+                    ('dvips was not able to process the following '
+                     'string:\n%s\n\n'
+                     'Here is the full report generated by dvips:\n%s '
+                     '\n\n' % (repr(tex.encode('unicode_escape')),
```
Same question.

tomspur commented:
Thanks for the review.

- The failing test is a flaky test.
- The unicode escaping was there before, so I left it in as is.
- I have a test around, but don't know how to use the `image_comparison` decorator and some pytest decorator to indicate an expected exception at the same time. I would prefer to continue looking into this, after this has been merged...

tacaswell commented:
Thanks for doing this tedious work!

Looking forward to the added test coverage.
